# Genome sizes - some draft notes

(c) Craig Duncan 2020


# A short summary:
Human genome: 3,257,319,537 bp ()
bp = 

 - A-T(DNA) or A-U(RNA); or
 - G-C (DNA and RNA)

Gbp-->
Wheat, barley, rye, pea?
| Species | Gbp | Name | Ensembl |
|:-----|:-----|:-----|:-----|
| H.Sapiens | 3.26 | Person | |
| H. vulgare |	4.79	| barley | |
| S. cereale |	6.67	| rye | |
| T. dicoccoides |	10.50 |	Emmer wheat | [link](http://plants.ensembl.org/Triticum_dicoccoides/Info/Index) |
| T. aestivum  |	14.50 | Common wheat | |

nb useful index of common names and Latin terms [https://www.ncbi.nlm.nih.gov/books/NBK208347/](https://www.ncbi.nlm.nih.gov/books/NBK208347/) and [here](https://www.inspection.gc.ca/plant-health/grains-and-field-crops/list-of-grains-and-field-crops/eng/1323244558875/1323244642996)

## Size:
RNA might only be a few thousand base pairs long (also, it has a ribose (sugar) backbone and is less stable than the 'de-oxyribose' DNA, especially in alkaline conditions; RNA gets attacked by enzymes)
DNA can be much longer (terminated for each chromosome e.g. centrimetres long for each, humans have 46 of them, roughly paired, in total about 2.3 metres long in every cell when unravelled!)

Function:
mRNA - messenger RNA copies (transcribes) portions of DNA (i.e. it is assembled at DNA sites) and takes them to the ribosome 'factories'.  Once mRNA has done its job, the DNA can be closed off again?

tRNA - brings in amino acids so the ribosome factories can make proteins (translation).  Think ''amino acid transfer', 'transport'

rRNA - the RNA that is part of the structure of the ribosome factories: initiates the protein building functions. i.e. you know this is localised to the mRNA.

https://www.technologynetworks.com/genomics/lists/what-are-the-key-differences-between-dna-and-rna-296719

# Genome size (wiki)

"Genome size is the total amount of DNA contained within one copy of a single genome. It is typically measured in terms of mass in picograms (trillionths (10−12) of a gram, abbreviated pg) or less frequently in Daltons or as the total number of nucleotide base pairs typically in megabases (millions of base pairs, abbreviated Mb or Mbp). One picogram equals 978 megabases.[1] In diploid organisms, genome size is used interchangeably with the term C-value. An organism's complexity is not directly proportional to its genome size; some single cell organisms have much more DNA than humans (see Junk DNA and C-value enigma). "

Genome size discussion : [monocot](https://www.biorxiv.org/content/10.1101/2020.06.01.126920v1)

# Definition

"Today, the term is used in 2 distinct ways to indicate either the total number of genes or the whole amount of nuclear DNA"
see (2012) Cytogenet Genome Res 2012;137:97–112
DOI: 10.1159/000338820

Big difference?

# Chromosomal patterns and assembly

Assembly Issues

 - Sequencing - entire genome is sequenced many times over
 - Assembly : attempt to assemble connected (contiguous) bases (for all chromosomes)
 - SNPs : For diploid organism, this is differences in single bases between the two chromosomes contributed by the parents (in gametes). *single nucleotide polymorphisms*. 
 - phased assembly: ability to identify gamete sources for all SNPs. 

Examples of differing chromosome sets.

Diploid organisms:
 - H. Sapiens
 - Drosopila

Haploids:
 - bee

Polyploidy:
 - Wheat

# Genome size intro

In any case genome size (GS) varies enormously, without apparent 'patterns' ... 

(are we looking at it from the macro first?  Then we see only the outcomes....)

"It has been found from data-mining GS databases
that GS is a useful cyto-taxonomical instrument at the level
of orders/superorders, providing genomic signatures char-
acterizing Monotremata, Marsupialia, Afrotheria, Xenarthra Laurasiatheria, and Euarchontoglires. A hypothetical ancestral mammalian-like GS of 2.9–3.7 pg has been suggested."


"For more than 50 years, we have been collecting data
on genome size (GS) and have discovered that the ge-
nomes of eukaryotic organisms vary over 200,000-fold in
size."

"When speaking of GS, the histori-
cally based consensus on the meaning of GS should be
clear: the total amount of a mature germ cell’s DNA. The
‘C’ terminology, introduced by Hewson Swift in 1950 in
an unfortunately ambiguous manner (since he used C to
refer to class, category, content, or constant in his first
paper on animal nuclei written for his PhD dissertation),
has been resolved thanks to a personal communication
by H. Swift himself to M.D. Bennett [Greilhuber et al.,
2005]: the term ‘C-value’ was intended to mean ‘constant’
(i.e. the amount of DNA that was characteristic of a par-
ticular genotype) and it was defined as ‘DNA content of
the unreplicated haploid chromosome complement’
[Bennett and Leitch, 2005]."

"In fact, the C-value
enigma conceptually better encapsulates all the relevant
questions related to GS variations. These questions in-
clude: what is the type of noncoding DNA that gives dif-
ferent genomes different sizes? What is the origin and the
intra- and intergenomic transfer of this noncoding DNA?
What, if any, are the functions (at any hierarchical orga-
nization level) that these DNA sequences provide?"

Summary :
A hypothetical ancestral mammalian-like GS of 2.9-3.7 pg (picograms)
Monotremata (∼2.97 pg) 
--->Afrotheria (∼5.5 pg) and Xenarthra (∼4.5 pg)
Marsupialia (∼4.07 pg), 
Euarchontoglires (∼3.4 pg)
Laurasiatheria (∼2.8 pg
(546 mammalian species sized from 5,488 living species) 

#  Plant genome sizes

[2020 monocot](https://www.biorxiv.org/content/10.1101/2020.06.01.126920v1)

"Any project seeking to deliver a plant or animal reference genome sequence must address the question as to the completeness of the assembly. "

Int J Mol Sci. 2018 Nov; 19(11): 3554.
Published online 2018 Nov 11. doi: 10.3390/ijms19113554
PMCID: PMC6274785
PMID: 30423889

https://www-ncbi-nlm-nih-gov.ezproxy.library.uwa.edu.au/pmc/articles/PMC6274785/

"Seventeen years have passed since the joint announcement of the human genome sequence [14,15]. This period has seen a number of attempts to complete the assembly, applying a variety of technologies [16,17]. All of these have reported a smaller genome size than what has, as of the end of 2017, been suggested in GRCh38.p12, the most recently released Genome Reference Consortium version, which comprises 3,257,319,537 bp. Assuming the Doležel et al. [18] conversion of 1 pg = 0.978 Gbp, 3.5 pg 1C DNA is equivalent to 3,423,000,000 bases. Thus, the 7 pg value represents an ~5.1% over-estimate of the GRCh38.p12 assembly prediction. This difference lies at the lower end of the error range predicted by Doležel and Greilhuber [13]. Given that the human reference genome is still incomplete, the expectation is that the gap between the 7 pg figure and the “real” human genome size will continue to diminish. Nevertheless, a 5% error is not dissimilar to the variation observed between estimates of nuclear DNA amounts of a given species produced by different laboratories [19,20]. Thus, the recommendation remains that the 7 pg figure continue to be used as the reference for measuring 2C values of both animal and plant genomes."
