# Notes on Genome sizes 

(c) Craig Duncan 2020-1. A work in progress.

Created 15 August 2020.  Last updated 8 February 2021.

# An illustration

Human genome: 3,257,319,537 bp

bp = 

 - A-T(DNA) or A-U(RNA); or
 - G-C (DNA and RNA)

Gbp-->
Wheat, barley, rye, pea?
| Species | Gbp | Name | Ensembl |
|:-----|:-----|:-----|:-----|
| A. Thaliana |0.125 | thale cress | [link](http://plants.ensembl.org/Arabidopsis_thaliana/Info/Index) |
| H.Sapiens | 3.26 | Person | |
| H. vulgare |	4.79	| barley | |
| S. cereale |	6.67	| rye | |
| T. dicoccoides |	10.50 |	Emmer wheat (Tetraploid) | [link](http://plants.ensembl.org/Triticum_dicoccoides/Info/Index) |
| T. aestivum  |	14.50 | Common wheat (Hexaploid)| [link](http://plants.ensembl.org/Triticum_aestivum/Info/Index) |
| P. polyphylla (PPY) | 82.55 | Paris | | 

nb useful index of common names and Latin terms [https://www.ncbi.nlm.nih.gov/books/NBK208347/](https://www.ncbi.nlm.nih.gov/books/NBK208347/) and [here](https://www.inspection.gc.ca/plant-health/grains-and-field-crops/list-of-grains-and-field-crops/eng/1323244558875/1323244642996)

The Arabidopsis Thaliana genome has haploid number of 5, and its somatic cell chromosomes are diploid, so there are 5 chromosome pairs (10 chromosomes) per cell.
Arabidopsis Genome Initiative. Analysis of the genome sequence of the flowering plant Arabidopsis thaliana. Nature. 2000 Dec;408(6814):796-815. DOI: 10.1038/35048692.

https://www.nature.com/articles/35048692.pdf
[Arabidopsis genome paper (2000)](https://www.nature.com/articles/35048692)

[Paris Polyphylla paper (2020)](https://www.biorxiv.org/content/10.1101/2020.06.01.126920v1)

# Relative sizes

Two units of measurement:

 - nt = nucleotides, which is essentially a length (a count of base pairs), and 
 - S = Sverdberg units where this refers to settling time in a centrifuge (so S units cannot be added together meaningfully)

## DNA

In prokaryotic organisms, the DNA molecules are present in the main cellular structure of the a single celled organism; in eurokaryotes there is a 'nuclear' membrane enclosing the genetic material in a subdivision of the cell, over which products must be transported: there is a division of both material and function.

DNA can be much longer than RNA (it is terminated for each chromosome e.g. 5 centrimetres long for each, humans have 46 of them, roughly paired into 23 'sets' of chromosome pairs).

It has been said that human DNA would equal a length of 2.3 metres long in every cell when unravelled.   An example, is [here](https://hypertextbook.com/facts/1998/StevenChen.shtml), where it is said that the average length of the 46 chromosomes in the human body is 5cm (46 x 5cm = 230cm = 2.3m).  In that calculation, it states that there are closer to 6 Gbp per cell (i.e. twice the human genome nt number).  The assumption here is that there a 2 x 23 chromosome *pairs* in every cell, and so the number of pairs to be laid out (>6Gbp) is twice as many as in the haploid genome per se.  

Does this make sense?

Number of base pairs (diploid genome) x distance between base pairs = total length 

e.g. (2 x 3 x 10^9) bp x 0.34 x 10^-9 (nanometres, nm) = 6 x 0.34 = 2.04m

Conventional wisdom [e.g. Dietzel lab](http://www.dietzellab.de/goodies/numbers.html) is that the base pairs are 0.34 nm apart (so closer to 3 nt per nanometre)

Number of base pairs?

The reference to base pairs in the 'human genome' is to the amount of genetic material needed to define the human species in 'haploid' (gamete) cells (like germ cells: eggs and sperm).  This is the minimum amount of sequential 'code' that describes the genetic material, but that's not actually how DNA is packed into most of the other, 'somatic' cells, which contain two copies of helical DNA, inherited from each of the parents, and this explains our chromosome 'pairs'.  So in somatic cells, there is 2 x 3.26Gb of genetic material and this is the total 'amount' of DNA that is commonly referred to when the length comparisons for ~2.3m are made. 

(2 x 3.26 x 10^9) bp x 0.34 x 10^-9 (nanometres, nm) = 6.52 x 0.34 = 2.22m

nb: A more recent (2019) academic discussion of DNA metrics is contained here:

Piovesan, A., Pelleri, M.C., Antonaros, F. et al. On the length, weight and GC content of the human genome. BMC Res Notes 12, 106 (2019). https://doi.org/10.1186/s13104-019-4137-z   [link](https://bmcresnotes.biomedcentral.com/articles/10.1186/s13104-019-4137-z)

Relevant data from that article:

|Base pairs (male) (Gbp) |Length(m)|Weight(picograms,pg)|Deduced gap (nanometres,nm)|
|---:|---:|---:|---:|
| 6.27|2.05|6.41|0.327|

It seems that a greater level of significant figures would be needed to ensure accuracy.

## RNA forms

mRNA - in eukaryotes, messenger RNA copies (transcribes) portions of DNA (i.e. it is assembled at DNA sites) and transported to the ribosome 'factories'.  Also, the mRNA must also have non-coding regions removed (spliced) in order to become 'mature'.  The size and length of mRNA varies depending on the genetic material that is copied into the mRNA coding sequence.  nb: 'positive sense' viral RNA is a strand of RNA that is already in the right orientation to be translated.

tRNA - brings in amino acids so the ribosome factories can make proteins (translation).  Think ''amino acid transfer', 'transport'

rRNA - the RNA that is part of the structure of the ribosome factories: initiates the protein building functions. i.e. you know this is localised to the mRNA.  Due to the persistency and consistency of rRNA across organisms, it is often used for evolutionary studies. 

## RNA substructures and sizes

RNA might only be a few thousand base pairs long (also, it has a ribose (sugar) backbone and is less stable than the 'de-oxyribose' DNA, especially in alkaline conditions; RNA gets attacked by enzymes).

Ribosomal RNA typically represents the largest portion of the genetic code representing an RNA sequence.  In the organisms's DNA sequence itself, the regions coding for the rRNA should be found in specific chromosome(s) and locations.  The codes may be repeated, and there may be several steps between initial translation of coded sequences (sometimes called 'precursor' regions) and the production of smaller units needed for biological activity.

In transcribed RNA, especially those rRNA involved in DNA transcription and protein building, the types are identifiable and generally predictable in size, but there are variable and persistent regions that alter amongst species.  They are often found tightly located within 'units' in the ribosomes. 

| Type | RNA SeqID | rRNA type | Species | Size (~) | Ref Link |
|:----------|:----------|:--------|:--------|:--------|:--------|
| eukaryotic | full | rRNA | human| 7216 nt | |
| eukaryotic | 5S | rRNA | human | 120 nt | [5S RNA database](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC99124/) |
| prokaryotic | 16S | rRNA | E. coli (*) | 1542 nt | [16S examples](https://www.ncbi.nlm.nih.gov/nuccore?term=33175%5BBioProject%5D+OR+33317%5BBioProject%5D) |
| prokaryotic | 16S | rRNA | H.pylori | 1503 nt | [16S NCBI](https://www.ncbi.nlm.nih.gov/nuccore/NR_044761.1) |
| eukaryotic | 18S | rRNA | human | 1869 nt | [18S5 NCBI link](https://www.ncbi.nlm.nih.gov/nuccore/NR_003286.2) |
| prokaryotic | 23S | rRNA | (bacteria) | 2904 nt | |
| eukaryotic | 28S | rRNA | human| 5025 nt | [28S DNA link](https://www.ncbi.nlm.nih.gov/gene?Cmd=DetailsSearch&Term=100008589) |

nb: 16S rRNA sequence differences were the basis of the proposal by Woese to introduce a distinction between prokaryotic bacteria and prokaryotic archea.  It has subsequently been suggested that archea may have diverged from gram-positive bacteria (single lipid layer in common):
Cavalier-Smith T (January 2002). "The neomuran origin of archaebacteria, the negibacterial root of the universal tree and bacterial megaclassification". International Journal of Systematic and Evolutionary Microbiology. 52 (Pt 1): 7–76. doi:10.1099/00207713-52-1-7. PMID 11837318.

Refs:

[Some indicative rRNA sizes across species](https://www.thermofisher.com/au/en/home/references/ambion-tech-support/rna-isolation/general-articles/ribosomal-rna-sizes.html)

[(2018)Epigenetic expression of 5S in A. Thaliana](https://pubmed.ncbi.nlm.nih.gov/29518237/) 

Quote: "A typical Arabidopsis 5S rRNA gene is 500 bp long, comprising a 120 bp transcribed sequence and a 380 bp spacer region."
i.e. the functional RNA that is transcribed is only 120bp, but the original DNA 'gene' within which this is contained is 500bp.

[PMC](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5887818/)

Variation among human 28S ribosomal RNA genes.  I L Gonzalez, J L Gorski, T J Campen, D J Dorney, J M Erickson, J E Sylvester, R D Schmickel
Proc Natl Acad Sci U S A. 1985 Nov; 82(22): 7666–7670. doi: 10.1073/pnas.82.22.7666
[Link](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC391394/)

[wet lab discussion 28S](https://www.researchgate.net/topic/28S-RNA)

Quote: "The rapid divergence rates of variable regions in the ribosomal gene may permit answers to the question of time of separation of closely related species."

Implications of sequence variation on the evolution of rRNA
Matthew M. Parks, Chad M. Kurylo, Jake E. Batchelder, C. Theresa Vincent, Scott C. Blanchard
Chromosome Res. 2019; 27(1): 89–93. Published online 2019 Feb 5. doi: 10.1007/s10577-018-09602-w
Correction in: Chromosome Res. 2019; 27(3): 285.

## Coronavirus

Compare above ribosomal RNA sizes to full SARS-Cov-2 genome:

| Type | RNA Sequence | rRNA type | Size (~) |
|:----------|:----------|:--------|:--------|
| coronavirus | full | RNA | 29903 nt |

i.e. 5 to 6 times larger than the very small intra-cellular protein-building apparatus.  Still 5 times smaller than phage T4 genome (see below).

[posting](https://virological.org/t/novel-2019-coronavirus-genome/319)

[library link](https://www.ncbi.nlm.nih.gov/nuccore/MN908947)

[sampling background](https://genomemedicine.biomedcentral.com/articles/10.1186/s13073-020-00751-4)

Virus stability issues:
2020, Wakida et al, Stability of RNA sequences derived from the coronavirus genome in
human cells, Biochemical and Biophysical Research Communications 527 (2020) 993e999
https://doi.org/10.1016/j.bbrc.2020.05.008

## Bacteriophages

| Name | Notes| Genome size | Researcher |
|:-----|:-----|:-----|:-----|
|T4 | Attacks E.Coli | 168,903 bp | Miller |
|A9 |  Bacteroidetes(?) | 735,000 bp | Al-Shayeb |

Miller et al. Bacteriophage T4 Genome.  MICROBIOLOGY AND MOLECULAR BIOLOGY REVIEWS, Mar. 2003, p. 86–156 1092-2172 DOI: 10.1128/MMBR.67.1.86–156.2003
[Link](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC150520/pdf/0005.pdf)

Miller's work suggests the origins of the T4 phage are ancient (before prokaryote and eukaryote division).

Megaphages:
(2020). Al-Shayeb et al. Clades of huge phages from across Earth's ecosystems, Nature (2020). DOI: 10.1038/s41586-020-2007-4 , https://nature.com/articles/s41586-020-2007-4  [Link](https://phys.org/news/2020-02-huge-bacteria-eating-viruses-gap-life.html)

nb - see extensive reproducibility information at end, including [Github link](https://github.com/rohansachdeva/assembly_repeats/blob/master/assembly_repeats.py)

# Genome size definitions

"Genome size is the total amount of DNA contained within one copy of a single genome. It is typically measured in terms of mass in picograms (trillionths (10−12) of a gram, abbreviated pg) or less frequently in Daltons or as the total number of nucleotide base pairs typically in megabases (millions of base pairs, abbreviated Mb or Mbp). One picogram equals 978 megabases.[1] In diploid organisms, genome size is used interchangeably with the term C-value. An organism's complexity is not directly proportional to its genome size; some single cell organisms have much more DNA than humans (see Junk DNA and C-value enigma). "

Genome size discussion : [P. Polyphylla](https://www.biorxiv.org/content/10.1101/2020.06.01.126920v1)

"Today, the term is used in 2 distinct ways to indicate either the total number of genes or the whole amount of nuclear DNA"
see (2012) Cytogenet Genome Res 2012;137:97–112
DOI: 10.1159/000338820

# Ploidy and genetic data quantity

Assembly Issues
 - Sequencing - entire genome is sequenced many times over
 - Assembly : attempt to assemble connected (contiguous) bases (for all chromosomes)
 - SNPs : For diploid organism, this is a reference to differences in single bases between the two chromosomes contributed by the parents (in gametes). *single nucleotide polymorphisms*. 
 - phased assembly: ability to identify gamete sources for all SNPs. 

The tendency or organisms to inherit more than one 'genome' sequences tell us how many replicated sequencies of DNA material we are likely to find.  The sequence in each chromosome is comparable (i.e. this explains SNPs) and so involves the potential to swap out genetic material at the same place in the sequence.   Bearing this in mind, there is a difference between:

(a) the size of the genome in haploid cells; and
(b) the genetic material based on the number of complete sets of chromosomes (i.e the carriers of the full genome) that appear in somatic cells ("ploidy").

This might create confusion in terminology when referring to 'DNA Content':

Suda, Jan & Krahulcová, Anna & Trávníček, Pavel & Krahulec, František. (2006). Ploidy Level versus DNA Ploidy Level: An Appeal for Consistent Terminology. TAXON. 55. 447-450. 10.2307/25065591. 

'Ploidy' itself is just a multiple of a single set of the chromosomes, expressed per cell.  e.g. 'Diploid' means two sets of chromosomes *per cell*.   In calculating the total number of chromosomes we can express this in terms of a single set of chromosomes (monoploid, often abbreviated to 'x'), or in terms of chromsomes in gametes (or haploid chromosome number 'n').  For adult organisms that are diploid, the 'x' and 'n' numbers are the same.

Some examples of counting chromosomes and the terminology used:

| Species | Ploidy | Chromosomes| Notes|
|:-----|:-----|:-----|:-----|
| *Common wheat*|Monoploid (x) | 7 | |
| |Haploid/gamete (n) |  21 | Triploid |
| |Somatic cells (2n)| 42|  Hexaploid |
| *Human*|Monoploid (x) | 23 | |
| |Haploid/gamete (n) |  23 | Monoploid |
| |Somatic cells (2n)| 46|  Diploid |

# Effect of aridity on genome sizes

A recent pre-print by Schley et al, titled 'The Ecology of Palm Genomes: Repeat-associated genome size expansion is constrained by aridity', suggests that in Palms, at least, the wetter environments may contribute to larger genome sizes than arid environments.

[Schley et al, 2021](https://www.biorxiv.org/content/10.1101/2021.11.04.467295v1)

# Wheat size and relative sizes for ploidy

Wheat is an interesting story in genetic evolution and hybridisation, since it contains 3 sets of chromosome pairs broadly identifiable with three ancient species and/or diploid genome progenitors, as follows:

|Name | Group Code | Ploidy | Chromosomes| Species|
|:-----|:-----|:-----|:-----|:-----|
| Einkorn Wheat|A(u)|Diploid (2n)|2x7=14| T.monococcum|
| Speltoids|B|Diploid (2n)|2x7=14| ~Aegilops speltoides(?)|
| Goat grass|D|Diploid (2n)|2x7=14| Aegilops Tauschii|

Wheat evolutionary history and size comparisons:

|Name | Genome Codes | Ploidy | Chromosomes| Species| Gbp |
|:-----|:-----|:-----|:-----|:-----|:-----|
| Einkorn Wheat|A(m)|Diploid (2n)|2x7=14| T.monococcum| 5.6 | --> ancestor of urartu. ('einkorn' = 'single grain')
| Red Wild Einkorn Wheat|A(u)|Diploid (2n)|2x7=14| T.urartu | 4.6 |
|||||||
| Speltoids|B|Diploid (2n)|2x7=14|~Aegilops speltoides(?)| |
|||||||
| Goat grass|D|Diploid (2n)|2x7=14| Aegilops Tauschii|4.0|
|||||||
| Emmer Wheat|BA(u)|Tetraploid (4n)|4x7=28| T.Dicoccoides|10.5 |
| Common Wheat|BA(u)D|Hexaploid (6n)|6x7=42| T.aestivum|14.5*| 

*possibly as high as 17Gbp based on recent research by International Wheat Genome Sequencing Consortium (IWGSC).*

There is some proportionality in the number of base pairs of common wheat, as compared to its tetraploid and diploid progenitors, where the relative amounts of DNA material are known from the progressive retention of more genomic material from the earlier species (A(u), B and D respectively).

Should even these stated base pairs for the DNA of each listed species be higher? Based on converting 1 chromosome of DNA equivalent weight into base pairs, and multiplying by the weight of the reference chromosome samples, there is reason to believe the actual number of base pairs in selected species of wheat are higher than historical assemblies shown above (as at 2018). 

"Any project seeking to deliver a plant or animal reference genome sequence must address the question as to the completeness of the assembly. "

"Seventeen years have passed since the joint announcement of the human genome sequence [14,15]. This period has seen a number of attempts to complete the assembly, applying a variety of technologies [16,17]. All of these have reported a smaller genome size than what has, as of the end of 2017, been suggested in GRCh38.p12, the most recently released Genome Reference Consortium version, which comprises 3,257,319,537 bp. Assuming the Doležel et al. [18] conversion of 1 pg = 0.978 Gbp, 3.5 pg 1C DNA is equivalent to 3,423,000,000 bases. Thus, the 7 pg value represents an ~5.1% over-estimate of the GRCh38.p12 assembly prediction. This difference lies at the lower end of the error range predicted by Doležel and Greilhuber [13]. Given that the human reference genome is still incomplete, the expectation is that the gap between the 7 pg figure and the “real” human genome size will continue to diminish. Nevertheless, a 5% error is not dissimilar to the variation observed between estimates of nuclear DNA amounts of a given species produced by different laboratories [19,20]. Thus, the recommendation remains that the 7 pg figure continue to be used as the reference for measuring 2C values of both animal and plant genomes."

Dolezel, Jaroslav & Janačížková, Janačížková & Simkova, Hana & Bartoš, Jan. (2018). One Major Challenge of Sequencing Large Plant Genomes Is to Know How Big They Really Are. International Journal of Molecular Sciences. 19. 19113554. 10.3390/ijms19113554. 

[link](https://www.researchgate.net/publication/328875915_One_Major_Challenge_of_Sequencing_Large_Plant_Genomes_Is_to_Know_How_Big_They_Really_Are)

## Triticum aestivum (Common Wheat)

This explanation about the evolutionary history of common wheat is also found at genome assembly library site enSemble: [enSembl](http://plants.ensembl.org/Triticum_aestivum/Info/Index).  Consistently, the relative sizes (base pairs) of each of the A, B and D genomic groups is proportionately smaller than the hybridised (and hexaploidy) common wheat.

The 'A' and 'B' genome groups came together to produce a tetraploid species, which only relatively recently (about 8,000 years) hybridised with the D genome group to form common wheat.  For this reason, the Ae.Tauschii genome (now mapped) and the D component of common wheat are more similar than the A and B genomes referable to those ancient species.

see Zimin et al (2017), The first near-complete assembly of the hexaploid bread wheat genome, Triticum aestivum, GigaScience, Volume 6, Issue 11, November 2017, gix097, [doi](https://doi.org/10.1093/gigascience/gix097)
URL: [Zimin link](https://academic.oup.com/gigascience/article/6/11/gix097/4561661) 

## Triticum urartu (Red Wild Einkorn Wheat)

[EnsemblUrartu](http://plants.ensembl.org/Triticum_urartu/Info/Index)

http://www.plosone.org/article/info%3Adoi%2F10.1371%2Fjournal.pone.0096855

## Triticum monococcum (Einkorn Wheat)

There is an excellent genomics resource on 'Einkorn Wheat' at the Jaiswal Lab at Oregan State University: URL [JaiswalLab](http://jaiswallab.cgrb.oregonstate.edu/genomics/wheat)

Fox SE, Geniza M, Hanumappa M, Naithani S, Sullivan C, et al. (2014) De Novo Transcriptome Assembly and Analyses of Gene Expression during Photomorphogenesis in Diploid Wheat Triticum monococcum. PLoS ONE 9(5): e96855. doi:10.1371/journal.pone.0096855

## Triricum dicoccum (Emmer Wheat)

Genomic resource : [EnsemblerEmmer](http://plants.ensembl.org/Triticum_dicoccoides/Info/Index)
(size/statistics): [EnsemblEmmerAnnot](http://plants.ensembl.org/Triticum_dicoccoides/Info/Annotation/#assembly)

## Aegilops tauschii (Tausch's goat grass)

genomic resource: [Nature](https://www.nature.com/articles/nature24486)
[doi](doi:10.1038/nature24486)

Ref:
Genome sequence of the progenitor of the wheat D genome Aegilops tauschii.
Ming-Cheng Luo, Yong Q. Gu, Daniela Puiu, Hao Wang, Sven O. Twardziok, Karin R. Deal, Naxin Huo, Tingting Zhu, Le Wang, Yi Wang et al. 2017. Nature. 551:498502.

# Comparative ploidy

Diploid organisms:
 - H. Sapiens. (i.e. 23 'paired' chromosomes = 6Gbp in total, but say only 3.26Gbp DNA haploid content needed to define the human 'genome')
 - Drosophila
 - Arabidopsis Thaliana 
 
 Ref:
Kaul, S., Koo, H. L., Jenkins, J., Rizzo, M., Rooney, T., Tallon, L. J., ... Somerville, C. (2000). Analysis of the genome sequence of the flowering plant Arabidopsis thaliana. Nature, 408(6814), 796-815. https://doi.org/10.1038/35048692 [Link](https://www.nature.com/articles/35048692)

Haploids:
 - bee (male drones)

Polyploids:
 - Wheat
 - Paris Polyphilla (tetraploid; some are diploid)

On the incidence of polyploidy in plants, see:

"Polyploidization is considered a major evolutionary force in plants."
Aversano R, Ercolano MR, Caruso I, Fasano C, Rosellini D, Carputo D. Molecular tools for exploring polyploid genomes in plants. Int J Mol Sci. 2012;13(8):10316-10335. doi:10.3390/ijms130810316


## Implications for size

Polyploidy doesn't increase the size of the 'genome': it increases the potential for SNPs when sampling DNA from cells, especially if it has been segmented first.

It usually increases cell size and organism size:
Ploidy and Size at Multiple Scales in the Arabidopsis Sepal
Dana O. Robinson, Jeremy E. Coate, Abhyudai Singh, Lilan Hong, Max Bush, Jeff J. Doyle, Adrienne H.K. Roeder
The Plant Cell Oct 2018, 30 (10) 2308-2329; DOI: 10.1105/tpc.18.00344

## Implications for sequence assembly

Current Strategies of Polyploid Plant Genome Sequence Assembly
Maria Kyriakidou, Helen H. Tai, Noelle L. Anglin, David Ellis, Martina V. Strömvik
Front Plant Sci. 2018; 9: 1660. Published online 2018 Nov 21. doi: 10.3389/fpls.2018.01660
PMCID: PMC6258962

Molecular Tools for Exploring Polyploid Genomes in Plants
Riccardo Aversano, Maria Raffaella Ercolano, Immacolata Caruso, Carlo Fasano, Daniele Rosellini, Domenico Carputo
Int J Mol Sci. 2012; 13(8): 10316–10335. Published online 2012 Aug 17. doi: 10.3390/ijms130810316

# Genome size benchmarking

In any case genome size (GS) varies enormously, without apparent 'patterns' .

## Mammalian

"It has been found from data-mining GS databases
that GS is a useful cyto-taxonomical instrument at the level
of orders/superorders, providing genomic signatures char-
acterizing Monotremata, Marsupialia, Afrotheria, Xenarthra Laurasiatheria, and Euarchontoglires. A hypothetical ancestral mammalian-like GS of 2.9–3.7 pg has been suggested."


"For more than 50 years, we have been collecting data
on genome size (GS) and have discovered that the ge-
nomes of eukaryotic organisms vary over 200,000-fold in
size."

"When speaking of GS, the histori-
cally based consensus on the meaning of GS should be
clear: the total amount of a mature germ cell’s DNA. The
‘C’ terminology, introduced by Hewson Swift in 1950 in
an unfortunately ambiguous manner (since he used C to
refer to class, category, content, or constant in his first
paper on animal nuclei written for his PhD dissertation),
has been resolved thanks to a personal communication
by H. Swift himself to M.D. Bennett [Greilhuber et al.,
2005]: the term ‘C-value’ was intended to mean ‘constant’
(i.e. the amount of DNA that was characteristic of a par-
ticular genotype) and it was defined as ‘DNA content of
the unreplicated haploid chromosome complement’
[Bennett and Leitch, 2005]."

"In fact, the C-value
enigma conceptually better encapsulates all the relevant
questions related to GS variations. These questions in-
clude: what is the type of noncoding DNA that gives dif-
ferent genomes different sizes? What is the origin and the
intra- and intergenomic transfer of this noncoding DNA?
What, if any, are the functions (at any hierarchical orga-
nization level) that these DNA sequences provide?"

Ref: 

Redi C, A, Capanna E: Genome Size Evolution: Sizing Mammalian Genomes. Cytogenet Genome Res 2012;137:97-112. doi: 10.1159/000338820

[link](https://www.karger.com/Article/Abstract/338820)

Summary :

A hypothetical ancestral mammalian-like GS of 2.9-3.7 pg (picograms)
Monotremata (∼2.97 pg) 
--->Afrotheria (∼5.5 pg) and Xenarthra (∼4.5 pg)
Marsupialia (∼4.07 pg), 
Euarchontoglires (∼3.4 pg)
Laurasiatheria (∼2.8 pg
(546 mammalian species sized from 5,488 living species) 

##  Plants

[P.Polyphilla](https://www.biorxiv.org/content/10.1101/2020.06.01.126920v1)


